core/src/qos_service.rs

```diff
@@ -177,28 +177,62 @@ impl QosService {
     pub fn update_or_remove_transaction_costs<'a>(
         transaction_costs: impl Iterator<Item = &'a TransactionCost>,
         transaction_qos_results: impl Iterator<Item = &'a transaction::Result<()>>,
-        retryable_transaction_indexes: &[usize],
+        transaction_committed_status: Option<&Vec<bool>>,
+        bank: &Arc<Bank>,
+    ) {
+        match transaction_committed_status {
+            Some(transaction_committed_status) => Self::update_transaction_costs(
+                transaction_costs,
+                transaction_qos_results,
+                transaction_committed_status,
+                bank,
+            ),
+            None => {
+                Self::remove_transaction_costs(transaction_costs, transaction_qos_results, bank)
+            }
+        }
+    }
+
+    fn update_transaction_costs<'a>(
+        transaction_costs: impl Iterator<Item = &'a TransactionCost>,
+        transaction_qos_results: impl Iterator<Item = &'a transaction::Result<()>>,
+        transaction_committed_status: &Vec<bool>,
         bank: &Arc<Bank>,
     ) {
         let mut cost_tracker = bank.write_cost_tracker().unwrap();
         transaction_costs
             .zip(transaction_qos_results)
-            .enumerate()
-            .for_each(|(index, (tx_cost, qos_inclusion_result))| {
-                // Only transactions that the qos service incuded have been added to the
-                // cost tracker.
-                if qos_inclusion_result.is_ok() && retryable_transaction_indexes.contains(&index) {
-                    cost_tracker.remove(tx_cost);
-                } else {
-                    // TODO: Update the cost tracker with the actual execution compute units.
-                    // Will have to plumb it in next; For now, keep estimated costs.
-                    //
-                    // let actual_execution_cost = 0;
-                    // cost_tracker.update_execution_cost(tx_cost, actual_execution_cost);
+            .zip(transaction_committed_status)
+            .for_each(|((tx_cost, qos_inclusion_result), was_committed)| {
+                // Only transactions that the qos service included have to be
+                // checked for update
+                if qos_inclusion_result.is_ok() {
+                    if *was_committed {
+                        cost_tracker.update_execution_cost(tx_cost, None);
```

Comment:
Can we leave this as a todo for now and follow up in a new PR when the actual cost is available here?

Reply:
Let's leave the piping work here, the next work #24311 is standing by as soon as this PR is merged.

Reply:
Why? This stuff is not related the issue that this PR is solving and it is slowing down review

Reply:
Yea, I see your point, and that's why it's 'None' there. 
For practical purpose, I just updated PR to address comments around the math, which is largely within `update_execution_cost()`, would be good to just push it through. 
But I am open to take this part of code into separate PR if that helps to merge this one quicker.

Reply:
Yeah please take it out to keep this PR focused on one functional change, thanks!

Reply:
Okie, removed

Reply:
Thanks!